refactor(spinner): migrate to Chakra and TypeScript

wanlingt opened the pull request:
- Convert Spinner component to use Chakra components and TypeScript
- AnswerSection component uses Chakra's SkeletonText directly

## Problem

Closes #273 

## Solution

Previously, 2 types of spinners were being used to display the loading state - one for pages and another for non-page sections (e.g. answer section, sending OTP)
These have been replaced with the Chakra spinner. For the answer section, Chakra's SkeletonText is used directly.


## Before & After Screenshots

**BEFORE**:

https://user-images.githubusercontent.com/56983748/133585462-4a283630-2281-4c9e-97e6-ce011d83c8aa.mov


https://user-images.githubusercontent.com/56983748/133585485-b1f53a1c-c96e-4c29-b612-508c72848352.mov




**AFTER**:



https://user-images.githubusercontent.com/56983748/133585788-20bd8abd-ff0d-4dd5-9727-63f169ed800d.mov



https://user-images.githubusercontent.com/56983748/133585621-fb1f08ea-fdca-48dc-a8a2-a7a684cbbe66.mov


<img src="https://user-images.githubusercontent.com/56983748/133721691-684d24e7-35b9-45ce-82c3-1a8a85167206.jpg" height="500px">

<img src="https://user-images.githubusercontent.com/56983748/133722268-5d552904-c740-4fe5-bc9a-900bb427c93b.jpg" height="500px">


## Tests

- Verify that pages use the updated Chakra spinners. These include
  - Landing page
  - Answer page
  - Header
  - Login page (sending OTP)
  - Edit form page
  - Post form page
  - Search results page

wanlingt committed:
refactor(spinner): migrate to Chakra and TypeScript

- convert Spinner component to use Chakra components and TypeScript
- change spinner in
AnswerSection component to skeleton

close #273

nauynix commented on client/src/components/Spinner/Spinner.component.tsx:
```diff
@@ -0,0 +1,22 @@
+  return type === 'paragraph' ? (
+    <SkeletonText />
+  ) : (
+    <div className="spinner" style={{ width: `${width}`, height: `${height}` }}>
```
Since we are migrating to chakra, the `div` should also be chakra and the properties are styled in this file, and thus we can remove the corresponding `scss` file.

mantariksh replied:
instead of wrapping in an extra `div` just for styling, why not just use the ChakraUI styling options?
```tsx
<PageSpinner w={width} h={height} />
```

wrapping in an extra div could also be slightly annoying for styling maintainability since the `Spinner` component returns two possible things with two different levels of DOM nesting

nauynix replied:
I believe we still have to wrap it in a `Center` or equivalent to center it.

wanlingt replied:
Yeah the alignment didn't work without `Center`, so I've left it in

mantariksh commented on client/src/components/Spinner/Spinner.component.tsx:
```diff
@@ -0,0 +1,22 @@
+import { SkeletonText, Spinner as PageSpinner } from '@chakra-ui/react'
+
+interface SpinnerProps {
+  type?: string
```
this isn't just any `string` right - you have a specific set of strings which are allowed (e.g. `'paragraph'`)? if so we can type this more strongly with [string literal types](https://www.typescriptlang.org/docs/handbook/literal-types.html#string-literal-types)

mantariksh commented on client/src/components/Spinner/Spinner.component.tsx:
```diff
@@ -0,0 +1,22 @@
+interface SpinnerProps {
+  type?: string
+  width: string | number
+  height: string | number
```
noticed that these are required types, yet in `AnswerSection.component.jsx`, they are not provided (and the code doesn't complain because the component isn't in TypeScript). perhaps these should be optional?

alternatively, consider this - it seems your goal here is to provide the consumer with two options:
1. render a `SkeletonText`, in which case styles need not be provided
2. render a `PageSpinner`, in which case the width and height must be provided

this approach allows you to specify the type more explicitly:
```ts
type SpinnerProps = 
  | { type: 'paragraph' }
  | { type: 'page'; width: string | number; height: string | number }
```

however, notice that the `type` key is redundant here. you can tell what the type is from the presence/absence of the width and height keys. this allows us to simplify further:
```ts
type SpinnerProps = 
  | Record<string, never>
  | { width: string | number; height: string | number }
```
where `Record<string, never>` indicates the absence of any props, and instead of `type === 'paragraph'`, you can check `width !== undefined`.

I know this probably seems SUPER nitpicky but a) hopefully it's useful for learning and b) strong type systems are built from solid basic types!

wanlingt replied:
I've gone with a combination of the 2 suggestions: 
1. Used SpinnerText directly in AnswerSection component
2. make `SpinnerProps` extend from ChakraUI's `SpinnerProps`

mantariksh reviewed (commented):
comments mostly for learning! not requesting changes bc mostly nitpicks, totally okay if this goes in as-is.

for your consideration - the `Spinner` component doesn't seem to be adding much abstraction over the ChakraUI `SkeletonText` and `PageSpinner` components. in fact, there are loads of props which consumers might want to pass to `SkeletonText` and `PageSpinner` (`width`, `height`, `speed`, `thickness` etc) but this component makes that impossible because it does not pass on any props to the ChakraUI components.

two possible solutions:
1. get rid of this component and just use `SkeletonText` and `PageSpinner` directly everywhere
2. make `SpinnerProps` extend from ChakraUI's `SpinnerProps` or `SkeletonProps` (not sure how easy this would be since `SpinnerProps` and `SkeletonProps` are different), and pass on the props internally to the relevant component

mantariksh commented:
also since our users are primarily on mobile, can we make it  a habit to include screenshots from mobile pls?

wanlingt committed:
refactor(spinner): separate Spinner and Skeleton Chakra components

- AnswerSection component uses SkeletonText directly from ChakraUI
- SpinnerProps extends from
ChakraUI's SpinnerProps

close #273

wanlingt commented:
> also since our users are primarily on mobile, can we make it a habit to include screenshots from mobile pls?

added!

wanlingt commented:
> also since our users are primarily on mobile, can we make it a habit to include screenshots from mobile pls?

Added!

mantariksh commented on client/src/components/Spinner/Spinner.component.tsx:
```diff
@@ -0,0 +1,38 @@
+import { Center, Spinner as PageSpinner } from '@chakra-ui/react'
+
+interface SpinnerProps {
```
actually ChakraUI exports all component props, which you can see from the [source](https://github.com/chakra-ui/chakra-ui/blob/main/packages/spinner/src/spinner.tsx), so you can just import `SpinnerProps`

mantariksh commented on client/src/components/Spinner/Spinner.component.tsx:
```diff
@@ -0,0 +1,38 @@
+import { Center, Spinner as PageSpinner } from '@chakra-ui/react'
```
nit: the name `PageSpinner` doesn't make its distinction from `Spinner` clear. I'd suggest `ChakraSpinner` to make it clear that `Spinner` is wrapping a ChakraUI component.

mantariksh commented on client/src/components/Spinner/Spinner.component.tsx:
```diff
@@ -0,0 +1,38 @@
+  speed,
+}: PageSpinnerProps): JSX.Element => {
+  return (
+    <Center className="spinner" h={height} w={width}>
```
would it work if we applied the width and height on the `Spinner` itself instead? that would allow you to spread the props so the whole component can look like this:
```ts
import {
  Center,
  Spinner as ChakraSpinner,
  SpinnerProps,
} from '@chakra-ui/react'

const Spinner = (props: SpinnerProps): JSX.Element => {
  return (
    <Center>
      <ChakraSpinner {...props} />
    </Center>
  )
}

export default Spinner
```

wanlingt replied:
Looks this if we apply `width` and `height` on `Spinner`, so probably not the best.

https://user-images.githubusercontent.com/56983748/133736244-d8d9b44a-7202-488a-98e6-1dd89f8b5982.mov



Without `width` and `height` applied to `Center`:

![Screenshot 2021-09-17 at 2 40 50 PM](https://user-images.githubusercontent.com/56983748/133736499-2e2ae716-e0c1-41fa-97f8-661c74ac224c.png)

With `height` applied to `Center`:

![Screenshot 2021-09-17 at 2 41 03 PM](https://user-images.githubusercontent.com/56983748/133736596-af9e5c0d-0b6f-40ed-adfd-f853dc40f22d.png)

LinHuiqing replied:
Not sure if this is a dumb question, but is there a need for a custom component file for the spinner? Would it make sense for us to use the Spinner component from Chakra directly? It looks very similar based on the docs: https://chakra-ui.com/docs/feedback/spinner

If we want customisations (which are shared across the site), we could extend the theme as well: https://chakra-ui.com/docs/theming/customize-theme#customizing-component-styles

mantariksh replied:
is it just a matter of adjusting the width and height to a more appropriate size?

wanlingt replied:
unfortunately no, that changes the size of the spinner but not its positioning

mantariksh commented on client/src/components/RichText/LinkControl.tsx:
```diff
@@ -90,7 +90,11 @@ export const LinkControl = ({
     <div className={styles.linkControl}>
       <div onClick={signalExpandShowModal} className="rdw-option-wrapper">
-        <BiLink size="18px" />
+        <Tooltip label="Link" placement="bottom" openDelay={500} fontSize="xs">
+          <span>
+            <BiLink size="18px" />
+          </span>
+        </Tooltip>
```
is this change meant to be in this PR? usually we try to scope PRs to a single set of changes, for a couple of reasons:
1. easier to review
2. if we need to revert commits, we can do so in isolation without affecting other changes

wanlingt committed:
refactor(spinner): center spinner using Chakra

close #273

LoneRifle commented on client/src/components/Spinner/Spinner.component.jsx:
```diff
@@ -1,15 +0,0 @@
-import React from 'react'
-import './Spinner.styles.scss'
-
-import { ReactComponent as PageSpinner } from '../../assets/PageSpinner.svg'
-import { ReactComponent as ComponentSpinner } from '../../assets/three-dots.svg'
```
you can take the opportunity to eliminate the two SVGs in this PR

wanlingt replied:
deleted

LoneRifle reviewed (approved):
lgtm, respond to comments then merge when ready

mantariksh reviewed (approved):
nice, the `centerWidth` and `centerHeight` solution is good because it still allows the ChakraUI inbuilt `width` and `height` to be applied to the Spinner if necessary

wanlingt committed:
refactor(spinner): remove spinner svgs from assets

close #273
